TAGradingServer/account/admin-latedays-exceptions.php

```diff
@@ -0,0 +1,752 @@
+	$student_table_db_data = retrieve_students_from_db($g_id);
+	$view->configure_table($student_table_db_data);
+	
+	//display
+	$view->display($state);
+	
+	/* END Process ---------------------------------------------------------- */
+
+} else {
+//User is NOT administrator -- operation not permitted.  Display error.
+
+    $view = new local_view();
+    $view->display('unauthorized');
+}
+
+include "../footer.php";
+exit;
+
+/* END MAIN ================================================================= */
+
+function parse_and_validate_csv($csv_file, &$data) {
+//IN:  * csv file name and path
+//     * (by reference) empty data array that will be filled.
+//OUT: TRUE should csv file be properly validated and data array filled.
+//     FALSE otherwise.
+//PURPOSE:  (1) vet uploaded csv file so it may be parsed.
+//          (2) create data array of csv information that may be batch upserted.
+	
+	//Validate file MIME type (needs to be "text/plain")
+	$file_info = finfo_open(FILEINFO_MIME_TYPE);
+	$mime_type = finfo_file($file_info, $_FILES['csv_upload']['tmp_name']);
+	finfo_close($file_info);
+	
+	if ($mime_type !== "text/plain") {
+		$data = null;
+		return false;
+	}
+	
+	$rows = file($csv_file, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
+	
+	if ($rows === false) {
+		$data = null;
+		return false;
+	}
+	
+	foreach($rows as $row) {
+	
+		$fields = explode(',', $row);
+		
+		//Each row has three fields
+		if (count($fields) !== 3) {
+			$data = null;
+			return false;
+		}
+		
+		//$fields[0]: Verify student exists in class (check by RCS ID)
+		if (!verify_student_in_db($fields[0])) {
+			$data = null;
+			return false;
+		}		
+		
+		//$fields[1]: Verify that gradeable exists (by title)
+		if (!identify_gradeable_in_db($fields[1])) {
+			$data = null;
+			return false;
+		}		
+		
+		//$fields[2]: Number of late day exceptions must be an integer >= 0
+		//            ctype_digit() returns false with negative integers as strings
+		if (!ctype_digit($fields[2])) {
+			$data = null;
+			return false;
+		}
+		
+		//Fields information seems okay.  Push fields onto data array.
+		$data[] = $fields;
+	}
+
+	//Validation successful.
+	return true;
+}
+
+/* END FUNCTION parse_and_validate_csv() ==================================== */
+
+function verify_student_in_db($student) {
+//IN:  RCS student ID
+//OUT: TRUE should RCS ID be found in the database
+//     FALSE otherwise
+//PURPOSE:  Verify that student is in database (indicating the student is enrolled)
+
+	//SQL for "old schema"
+	$sql = <<<SQL
+SELECT
+	COUNT(1)
+FROM
+	students
+WHERE
+	student_rcs=?
+SQL;
+
+	//SQL for "new schema"
+	//FIXME: update 'user_group' property to match the STUDENTS group
+/* DISABLED CODE ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+	$sql = <<<SQL
+SELECT
+	COUNT(1)
+FROM users
+WHERE
+	user_id=?
+AND
+	user_group=0
+SQL;
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ END DISABLED CODE */
+
+	\lib\Database::beginTransaction();
```

Comment:
Remove this line and the one for commit. There's no reason to wrap a single SQL statement (especially if it's a select) in a transaction as there's no way to leave the database in a broken state.